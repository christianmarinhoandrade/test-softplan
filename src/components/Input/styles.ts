import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  margin-bottom: 1rem;
  label: {
    font-size: 1rem;
    line-height: 1;
    padding-bottom: 0.5rem;
  }
`
type InputProps = {
  error: boolean
}

export const Input = styled.input<InputProps>`
  ${({ theme, error, disabled }) => css`
    border: 1px solid ${error ? 'red' : '#eee'};
    display: block;
    width: 100%;
    font-size: 1rem;
    padding: 0.8rem;
    border-radius: 0.4rem;
    background-color: #eee;
    transition: 0.2s;

    ${
      disabled &&
      css`
    &:disabled {
      opacity: 0.5;
      `
    }}

    ${
      !disabled &&
      css`
        & :focus,
        :hover {
          outline: none;
          background: #fff;
          border-color: ${theme.lightBlue};
        }
      `
    }
  `}
`
export const Error = styled.p`
  color: red;
  font-size: 0.875rem;
  margin-top: 0.25rem;
`
