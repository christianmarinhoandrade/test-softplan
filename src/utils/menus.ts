export const getMenu = (
) => {
  return [
    {
      title: 'teste',
      path: '/teste',
      name: 'agendamento'
    }
  ]
}
