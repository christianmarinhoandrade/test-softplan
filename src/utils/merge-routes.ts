import { RouteProps } from 'react-router-dom'
const routesContext = require.context('../pages', true, /.routes./)

export type RouteMergeProps = {
  name?: string
  path?: string
} & RouteProps

const routes: RouteMergeProps[] = []

routesContext.keys().forEach((route) => {
  routes.push(...routesContext(route).default)
})

export default routes
