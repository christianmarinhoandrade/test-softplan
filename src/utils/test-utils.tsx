import { ReactElement } from 'react'
import { render } from '@testing-library/react'

import { Provider as ProviderRedux } from 'react-redux'
import store from 'store'

import ThemeProvider from 'providers/theme'

const customRender = (ui: ReactElement) =>
  render(
    <ProviderRedux store={store}>
      <ThemeProvider>{ui}</ThemeProvider>
    </ProviderRedux>
  )

export * from '@testing-library/react'
export { customRender as render }
