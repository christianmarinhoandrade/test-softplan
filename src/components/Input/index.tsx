/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { forwardRef, InputHTMLAttributes, Ref } from 'react'
import { DeepMap, FieldError } from 'react-hook-form'

import * as S from './styles'

type InputProps = {
  label: string
  errors?: DeepMap<any, FieldError>
}

const Input = (
  {
    label,
    id,
    errors,
    ...props
  }: InputProps & InputHTMLAttributes<HTMLInputElement>,
  ref: Ref<HTMLInputElement>
) => {
  return (
    <S.Wrapper>
      <label htmlFor={id}>{label}</label>
      <S.Input
        id={id}
        name={id}
        ref={ref}
        error={!!(errors && errors[id!]?.message)}
        {...props}
      />
      {errors && errors[id!]?.message && (
        <S.Error>{errors[id!]?.message}</S.Error>
      )}
    </S.Wrapper>
  )
}

export default forwardRef(Input)
