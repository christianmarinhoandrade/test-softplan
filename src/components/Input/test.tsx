import { render, screen, waitFor } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'

import Input from '.'

describe('<Input />', () => {
  it('Renders with Label', () => {
    render(<Input label="Label" id="Label" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Changes its value when typing', async () => {
    const onInputChange = jest.fn()
    render(<Input onChange={onInputChange} label="TextField" id="TextField" />)

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInputChange).toHaveBeenCalledTimes(text.length)
    })
  })

  it('Does not changes its value when disabled', async () => {
    const onInputChange = jest.fn()
    render(
      <Input
        onChange={onInputChange}
        label="TextField"
        id="TextField"
        disabled
      />
    )

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()

    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onInputChange).not.toHaveBeenCalled()
  })

  it('Renders with error', () => {
    const { container } = render(
      <Input
        label="TextField"
        id={'textfield'}
        errors={{ textfield: { message: 'Error message' } }}
      />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Is accessible by tab', () => {
    render(<Input label="TextField" id="TextField" />)

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('Is not accessible by tab when disabled', () => {
    render(<Input label="TextField" id="TextField" disabled />)

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).not.toHaveFocus()
  })
})
