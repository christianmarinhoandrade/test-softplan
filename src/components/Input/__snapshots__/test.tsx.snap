// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> Renders with error 1`] = `
.c0 {
  margin-bottom: 1rem;
}

.c0 label: {
  font-size: 1rem;
  line-height: 1;
  padding-bottom: 0.5rem;
}

.c1 {
  border: 1px solid red;
  display: block;
  width: 100%;
  font-size: 1rem;
  padding: 0.8rem;
  border-radius: 0.4rem;
  background-color: #eee;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c1 :focus,
.c1:hover {
  outline: none;
  background: #fff;
  border-color: #3b97de;
}

.c2 {
  color: red;
  font-size: 0.875rem;
  margin-top: 0.25rem;
}

<div
  class="c0"
>
  <label
    for="textfield"
  >
    TextField
  </label>
  <input
    class="c1"
    id="textfield"
    name="textfield"
  />
  <p
    class="c2"
  >
    Error message
  </p>
</div>
`;
